File: analytics/src/test/java/io/appmetrica/analytics/impl/component/clients/ReporterClientUnitFactoryTest.java
package io.appmetrica.analytics.impl.component.clients;

import io.appmetrica.analytics.impl.component.ComponentId;
import io.appmetrica.analytics.impl.component.DispatcherComponentFactory;
import io.appmetrica.analytics.impl.component.RegularDispatcherComponent;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ReporterClientUnitFactoryTest extends ClientUnitFactoryBaseTest {

    @Before
    public void setUp() {
        super.setUp();
        when(mComponentsRepository
                .getOrCreateRegularComponent(any(ComponentId.class), same(mCommonArguments), any(DispatcherComponentFactory.class))
        ).thenReturn(mock(RegularDispatcherComponent.class));
    }

    @Test
    public void testCreateClientUnit() {
        super.testCreateClientUnit(RegularClientUnit.class, ComponentId.class);
    }

    @Override
    protected ClientUnitFactory createClientUnitFactory() {
        return new ReporterClientUnitFactory();
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/component/clients/ReporterClientUnitFactoryTest.kt
package io.appmetrica.analytics.impl.component.clients

import android.content.Context
import io.appmetrica.analytics.impl.component.CommonArguments
import io.appmetrica.analytics.impl.component.ComponentId
import io.appmetrica.analytics.impl.component.RegularDispatcherComponent
import io.appmetrica.analytics.impl.component.RegularDispatcherComponentFactory
import io.appmetrica.analytics.impl.component.ReporterComponentUnit
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.constructionRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify

internal class ReporterClientUnitFactoryTest : CommonTest() {

    private val componentUnitFactory: ComponentUnitFactory<ReporterComponentUnit> = mock()
    private val context: Context = mock()

    private val packageName = "package name"
    private val apiKey = "api key"

    private val clientDescription: ClientDescription = mock {
        on { packageName } doReturn packageName
        on { apiKey } doReturn apiKey
    }

    private val sdkConfig: CommonArguments = mock()
    private val regularDispatcherComponent: RegularDispatcherComponent<*> = mock()

    private val repository: ComponentsRepository = mock {
        on { getOrCreateRegularComponent(any(), any(), any()) } doReturn regularDispatcherComponent
    }

    @get:Rule
    val componentIdRule = constructionRule<ComponentId>()
    private val componentId by componentIdRule

    @get:Rule
    val regularClientUnit = constructionRule<RegularClientUnit>()

    @get:Rule
    val regularDispatcherComponentFactoryRule = constructionRule<RegularDispatcherComponentFactory<*>>()
    private val regularDispatcherComponentFactory by regularDispatcherComponentFactoryRule

    private val factory by setUp { ReporterClientUnitFactory(componentUnitFactory) }

    @Test
    fun createClientUnit() {
        assertThat(factory.createClientUnit(context, repository, clientDescription, sdkConfig))
            .isEqualTo(regularClientUnit.constructionMock.constructed().first())
        assertThat(regularClientUnit.constructionMock.constructed()).hasSize(1)
        assertThat(regularClientUnit.argumentInterceptor.flatArguments())
            .containsExactly(context, regularDispatcherComponent)

        verify(repository).getOrCreateRegularComponent(componentId, sdkConfig, regularDispatcherComponentFactory)

        assertThat(componentIdRule.constructionMock.constructed()).hasSize(1)
        assertThat(componentIdRule.argumentInterceptor.flatArguments()).containsExactly(packageName, apiKey)

        assertThat(regularDispatcherComponentFactoryRule.constructionMock.constructed()).hasSize(1)
        assertThat(regularDispatcherComponentFactoryRule.argumentInterceptor.flatArguments())
            .containsExactly(componentUnitFactory)
    }
}
